tests/integration/relations/opensearch_provider/application-charm/src/charm.py

```diff
@@ -0,0 +1,239 @@
+
+        username = databag.get("username")
+        password = databag.get("password")
+        host = databag.get("endpoints").split(",")[0]
+        host_addr = host.split(":")[0]
+        port = host.split(":")[1]
+
+        logger.info(f"sending {method} request to {endpoint}")
+        try:
+            response = self.request(method, endpoint, port, username, password, host_addr, payload)
+        except OpenSearchHttpError as e:
+            response = [str(e)]
+        logger.info(response)
+
+        event.set_results({"results": json.dumps(response)})
+
+    # =================================
+    #  Opensearch connection functions
+    # =================================
+
+    def relation_request(
+        self,
+        relation_id: int,
+        method: str,
+        endpoint: str,
+        payload: Optional[Dict[str, any]] = None,
+    ) -> Union[Dict[str, any], List[any]]:
+        """Make an HTTP request to a specific relation."""
+        databag = self.first_database.fetch_relation_data()[relation_id]
+        logging.error(databag)
+        username = databag.get("username")
+        password = databag.get("password")
+        endpoints = databag.get("endpoints", "").split(",")
+
+        if None in [username, password] or len(endpoints) == 0:
+            raise OpenSearchHttpError
+
+        host = endpoints[0].split(":")[0]
+        port = int(endpoints[0].split(":")[1])
+
+        return self.request(
+            method,
+            endpoint,
+            port,
+            username,
+            password,
+            host,
+            payload=payload,
+        )
+
+    def request(
+        self,
+        method: str,
+        endpoint: str,
+        port: int,
+        username: str,
+        password: str,
+        host: str,
+        payload: Optional[Dict[str, any]] = None,
+    ) -> Union[Dict[str, any], List[any]]:
+        """Make an HTTP request.
+
+        TODO swap this over to a more normal opensearch client
+
+        Args:
+            method: matching the known http methods.
+            endpoint: relative to the base uri.
+            payload: JSON / map body payload.
+            host: host of the node we wish to make a request on.
+            port: the port for the server.
+            username: the username to use for authentication
+            password: the password for {username}
+        """
+        if None in [endpoint, method]:
+            raise ValueError("endpoint or method missing")
+
+        if endpoint.startswith("/"):
+            endpoint = endpoint[1:]
+
+        full_url = f"https://{host}:{port}/{endpoint}"
+
+        request_kwargs = {
+            "verify": False,  # TODO this should be a path to a cert once this relation has TLS.
+            "method": method.upper(),
+            "url": full_url,
+            "headers": {"Content-Type": "application/json", "Accept": "application/json"},
+        }
+
+        if payload:
+            request_kwargs["data"] = payload
```

Comment:
I think this should have `json.dumps(payload)`, right?

Reply:
I've updated this to add it as-is if it's a string, and json.dumps if it's a dict.